libraries/standard/mqtt/utest/mqtt_lightweight_utest.c

```diff
@@ -235,6 +323,1154 @@ static size_t encodeString( uint8_t * pDestination,
+                                            subscriptionCount,
+                                            &remainingLength,
+                                            NULL );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+
+    /* Verify empty subscription list fails.  */
+    memset( ( void * ) &subscriptionList, 0x0, sizeof( subscriptionList ) );
+    subscriptionCount = 0;
+    status = MQTT_GetUnsubscribePacketSize( &subscriptionList,
+                                            subscriptionCount,
+                                            &remainingLength,
+                                            &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    /* Verify good case. */
+    memset( ( void * ) &subscriptionList, 0x0, sizeof( subscriptionList ) );
+    subscriptionList.qos = MQTTQoS0;
+    subscriptionList.pTopicFilter = "/example/topic";
+    subscriptionList.topicFilterLength = sizeof( "/example/topic" );
+    subscriptionCount = sizeof( subscriptionList ) / sizeof( MQTTSubscribeInfo_t );
+    status = MQTT_GetUnsubscribePacketSize( &subscriptionList,
+                                            subscriptionCount,
+                                            &remainingLength,
+                                            &packetSize );
+    TEST_ASSERT_EQUAL_INT( MQTTSuccess, status );
+    TEST_ASSERT_GREATER_THAN( remainingLength, packetSize );
+}
+
+/**
+ * @brief Tests that MQTT_SerializeSubscribe works as intended.
+ */
+void test_MQTT_SerializeSubscribe( void )
+{
+    MQTTSubscribeInfo_t subscriptionList;
+    size_t subscriptionCount = 1;
+    size_t remainingLength = 0;
+    uint16_t packetIdentifier;
+    uint8_t buffer[ 25 ];
+    size_t bufferSize = sizeof( buffer );
+    size_t packetSize = bufferSize;
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTFixedBuffer_t fixedBuffer = { .pBuffer = buffer, .size = bufferSize };
+
+    const uint16_t PACKET_ID = 1;
+
+    /* Verify bad parameters fail. */
+    status = MQTT_SerializeSubscribe( NULL,
+                                      subscriptionCount,
+                                      PACKET_ID,
+                                      remainingLength,
+                                      &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    status = MQTT_SerializeSubscribe( &subscriptionList,
+                                      subscriptionCount,
+                                      0,
+                                      remainingLength,
+                                      &fixedBuffer );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
+
+    status = MQTT_SerializeSubscribe( &subscriptionList,
+                                      subscriptionCount,
+                                      PACKET_ID,
+                                      remainingLength,
+                                      NULL );
+    TEST_ASSERT_EQUAL_INT( MQTTBadParameter, status );
```

Comment:
Don't we have 2 more paths to cover for invalid params? subscriptionCount == 0U and packetSize > pBuffer->size

Reply:
subscriptionCount == 0 is handled in line 704. packetSize > pBuffer->size is handled in line 726, though I think that's only because the buffer is too small. We can make that a separate test where we more explicitly set the buffer size to a low value